### ISSUE TITLE
Process.run incorrect arg quoting on Windows

### ISSUE THREAD

**Issue by kjk** [dart-lang/sdk#42571]
```
> dart --version
Dart VM version: 2.9.0-19.0.dev.flutter-63cf56d925 (be) (Fri Jun 26 01:56:03 2020 +0000) on "windows_x64"
```

Run the following from Windows cmd-line (`cmd.exe` or `powershell`):
```
> start "https://blog.kowalczyk.info/?foo=bar&al=me"
```

This opens default web browser with url `https://blog.kowalczyk.info/?foo=bar&al=me`

This Dart program should do the same:
```dart
import 'dart:io';

final url = "https://blog.kowalczyk.info/?foo=bar&me=him";

Future<ProcessResult> launchURL(String url) {
  stdout.writeln("launchURL: " + url);
  return Process.run("start", [url], runInShell: true);
}

void main() {
  launchURL(url);
}
```

(name it `main.dart` and run with: `dart run main.dart`)

This opens a web browser with url:
* `https://blog.kowalczyk.info/?foo=bar` 
instead of:
* `https://blog.kowalczyk.info/?foo=bar&me=him`

My guess is that `Process.run` incorrectly escapes/quotes the cmd-line argument. If the argument has `&` in it, it should be quoted with `"`.

I found it while trying to do `github_graphql_client` Flutter code lab on Windows. 
It uses `package:url_launcher/url_launcher.dart` which doesn't support Windows so I tried to use `Process.run`. Unfortunately, this bug breaks GitHub OAuth callback (because the url has `&` in it and everything after `&` is stripped`)

**Comment by kjk** [comment]
It looks like the code is https://github.com/dart-lang/sdk/blob/master/runtime/bin/process_win.cc#L419 and it doesn't seem to do any escaping.

For reference, you can copy the implementation from Go: https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/syscall/exec_windows.go#L86

**Comment by zichangg** [comment]
@kjk Thanks for posting.
The problem is indeed what you said. The url should be enclosed within double quotes.
I'm not familiar with Go but [one thing](https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/syscall/exec_windows.go#L90) looks suspicious is that it appends a space to all non-empty arg. It will make sure urls get surrounded by double quotes.

what if I do `start "https://blog.kowalczyk.info/?foo=bar&me=him" & dir`? The second ampersand will become `"& "`, if I understand the code correctly. Is Go able to get the correct result?

**Comment by kjk** [comment]
The thing is that Go doesn't have equivalent of `runInShell` option.

So I believe quoting rules should be:
* if `runInShell` is false, construct the arg string so that when parsed using algorithm described in https://docs.microsoft.com/en-us/previous-versions/ms880421(v=msdn.10), it would produce the original array of args, as is done in Go
* if `runInShell` is true, if arg contains `&`, the `&` should be escaped with `^`. This is because with `runInShell` the command is (should be) executed as: `cmd.exe /c "start https://blog.kowalczyk.info/?foo=bar^&me=him"` (notice that `&` is escaped with `^`, not `\`).  This is because `&` is meaningful to `cmd.exe`

As to go quoting, it does the right thing.
Given args `a` and `b`, it constructs `a b`. Without that space between non-empty args, it would be `ab`.

**Comment by zichangg** [comment]
@kjk We had escaping in the dart code. Forgot to link in the previous comment. https://github.com/dart-lang/sdk/blob/b4c60e6e1c6b58fe96d6f19fba20b8f2844af5fc/sdk/lib/_internal/vm/bin/process_patch.dart#L330

I tested with python 2.7. It also has the same problem. You can run the command below:
```python
subprocess.Popen(['start', url], shell=True)
```

The command passed down should be `cmd.exe /c "start https://blog.kowalczyk.info/?foo=bar&me=him"`. ( double quotes surrounded url) instead of `cmd.exe /c start https://blog.kowalczyk.info/?foo=bar&me=him`. I believe this is a corner case that is not covered by Windows parsing rules.

> if runInShell is true, if arg contains &, the & should be escaped with ^. This is because with runInShell the command is (should be) executed as: cmd.exe /c "start https://blog.kowalczyk.info/?foo=bar^&me=him" (notice that & is escaped with ^, not \). This is because & is meaningful to cmd.exe

I doubt that will work for all cases. What about `start "https://blog.kowalczyk.info/?foo=bar&me=him" & dir`? The `&` before `dir` should not be escaped.
```dart
Process.run("start", [url, '&', 'dir'], runInShell: true);
```

**Comment by kjk** [comment]
Fair enough if it's considered out of scope as it can be done in user code.

The fact that if you use `runInShell` on Windows you should escape `&` with `^&` in args could be a doc addition. 

It really stumped me. I only figured it out when doing research for this bug report.

**Comment by a-siva** [comment]
@kwalrath looks like we need some additional documentation in `Process.run `around escaping certain characters on windows (see [https://ss64.com/nt/syntax-esc.html](https://ss64.com/nt/syntax-esc.html)). Can you suggest how we should incorporate that into the API documentation.

**Comment by kwalrath** [comment]
Looking at https://api.dart.dev/be/dart-io/Process/run.html... I think I'd add a platform note with a brief description and link to more info. Maybe add some example code, especially if you can show both *nix and Windows code. The note could be formatted specially (**Windows note:**) or could just be in a separate paragraph (maybe starting with a bold  "**On Windows**, ..." so people can't miss it).
